ddl/index.go

```diff
@@ -532,138 +549,100 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
+	workerCnt := defaultWorkers
 	taskOpInfo := &indexTaskOpInfo{
-		tblIndex:  tables.NewIndex(t.Meta(), indexInfo),
-		colMap:    colMap,
-		nextCh:    make(chan int64, 1),
-		taskRetCh: make(chan *taskResult, taskCnt),
+		tblIndex: tables.NewIndex(t.Meta(), indexInfo),
+		colMap:   colMap,
 	}
 
+	taskBatch := int64(defaultTaskHandleCnt)
 	addedCount := job.GetRowCount()
-	taskStartHandle := reorgInfo.Handle
+	// The previous done handle is used to be recorded in the reorg info.
+	prevDoneHandle, baseHandle := reorgInfo.Handle, reorgInfo.Handle
 
+	workers := make([]*worker, workerCnt)
+	for i := 0; i < workerCnt; i++ {
+		ctx := d.newContext()
+		workers[i] = newWorker(ctx, i, int(taskBatch), len(cols), len(taskOpInfo.colMap))
+	}
 	for {
 		startTime := time.Now()
 		wg := sync.WaitGroup{}
-		for i := 0; i < taskCnt; i++ {
+		for i := 0; i < workerCnt; i++ {
 			wg.Add(1)
-			go d.doBackfillIndexTask(t, taskOpInfo, taskStartHandle, &wg)
-			doneHandle := <-taskOpInfo.nextCh
-			// There is no data to seek.
-			if doneHandle == taskStartHandle {
-				break
-			}
-			taskStartHandle = doneHandle + 1
+			workers[i].setTaskNewRange(baseHandle, baseHandle+taskBatch)
+			// TODO: Consider one worker to one goroutine.
+			go workers[i].doBackfillIndexTask(t, taskOpInfo, &wg)
+			baseHandle += taskBatch
 		}
 		wg.Wait()
 
-		retCnt := len(taskOpInfo.taskRetCh)
-		taskAddedCount, doneHandle, err := getCountAndHandle(taskOpInfo)
-		// Update the reorg handle that has been processed.
-		if taskAddedCount != 0 {
-			err1 := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
-				return errors.Trace(reorgInfo.UpdateHandle(txn, doneHandle+1))
-			})
-			if err1 != nil {
-				if err == nil {
-					err = err1
-				} else {
-					log.Warnf("[ddl] add index failed when update handle %d, err %v", doneHandle, err)
-				}
-			}
-		}
-
+		taskAddedCount, doneHandle, isEnd, err := getCountAndHandle(prevDoneHandle, workers)
 		addedCount += int64(taskAddedCount)
 		sub := time.Since(startTime).Seconds()
+		if err == nil {
+			err = d.isReorgRunnable()
+		}
 		if err != nil {
-			log.Warnf("[ddl] total added index for %d rows, this task add index for %d failed, take time %v",
-				addedCount, taskAddedCount, sub)
+			// Update the reorg handle that has been processed.
+			err1 := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
+				return errors.Trace(reorgInfo.UpdateHandle(txn, doneHandle))
+			})
+			log.Warnf("[ddl] total added index for %d rows, this task add index for %d failed, take time %v, update handle err %v",
+				addedCount, taskAddedCount, sub, err1)
 			return errors.Trace(err)
 		}
 		d.setReorgRowCount(addedCount)
 		batchHandleDataHistogram.WithLabelValues(batchAddIdx).Observe(sub)
 		log.Infof("[ddl] total added index for %d rows, this task added index for %d rows, take time %v",
 			addedCount, taskAddedCount, sub)
 
-		if retCnt < taskCnt {
+		prevDoneHandle, baseHandle = doneHandle, doneHandle+1
```

Comment:
We don't need to set `baseHandle` to `doneHandle+1`, it is already the value we need.

Reply:
These handles may be deleted many in bulk,  so the `doneHandle + 1` may be much larger than `baseHandle`.